core/templates/dev/head/pages/admin/roles_tab/RolesTabDirective.js

```diff
@@ -0,0 +1,129 @@
+        $scope.UPDATABLE_ROLES = GLOBALS.UPDATABLE_ROLES;
+        $scope.VIEWABLE_ROLES = GLOBALS.VIEWABLE_ROLES;
+        $scope.graphData = GLOBALS.ROLE_GRAPH_DATA;
+        $scope.showResultRoles = false;
+        $scope.result = {};
+        $scope.setStatusMessage('');
+        $scope.showUpdateForm = ADMIN_SHOW_UPDATE_ROLE;
+
+        // Calculating initStateId and finalStateIds for graphData
+        // Since role graph is acyclic, node with no incoming edge
+        // is initState and nodes with no outgoing edge are finalStates.
+        var hasIncomingEdge = [];
+        var hasOutgoingEdge = [];
+        for (var i = 0; i < $scope.graphData.links.length; i++) {
+          hasIncomingEdge.push($scope.graphData.links[i].target);
+          hasOutgoingEdge.push($scope.graphData.links[i].source);
+        }
+        var finalStateIds = [];
+        for (var i in $scope.graphData.nodes) {
+          if ($scope.graphData.nodes.hasOwnProperty(i)) {
+            if (hasIncomingEdge.indexOf(i) == -1) {
+              $scope.graphData.initStateId = i;
+            }
+            if (hasOutgoingEdge.indexOf(i) == -1) {
+              finalStateIds.push(i);
+            }
+          }
+        }
+        $scope.graphData.finalStateIds = finalStateIds;
+
+        $scope.submitRoleViewForm = function(values) {
+          if (AdminTaskManagerService.isTaskRunning()) {
+            return;
+          }
+
+          $scope.setStatusMessage('Processing query...');
+
+          AdminTaskManagerService.startTask();
+          $scope.result = {};
+          $http.get(ADMIN_ROLE_HANDLER_URL, {
+            params: {
+              method: values.method,
+              role: values.role,
+              username: values.username
+            }
+          }).then(function(response) {
+            $scope.result = response.data;
+            if (Object.keys($scope.result).length == 0) {
+              $scope.showResultRoles = false;
+              $scope.setStatusMessage('No results.');
+            }
+            else {
+              $scope.showResultRoles = true;
+              $scope.setStatusMessage('Success.');
+            }
+          }, function(errorResponse) {
+            $scope.setStatusMessage(
+              'Server error: ' + errorResponse.data.error);
+          });
+          if($scope.viewFormValues.username) {
+            $scope.viewFormValues.username = '';
+          }
+          if($scope.viewFormValues.role) {
+            $scope.viewFormValues.role = '';
+          }
+          AdminTaskManagerService.finishTask();
+        }
+
+        $scope.submitUpdateRoleForm = function(values) {
+          if (AdminTaskManagerService.isTaskRunning()) {
+            return;
+          }
+          $scope.form.updateRoleForm.$setPristine();
+          // if($scope.updateFormValues.username) {
+          //   $scope.updateFormValues.username = '';
+          // }
+          //$scope.updateFormValues.newRole = '';
+          
```

Comment:
Both `$scope.updateRoleForm` and `$scope.viewRoleForm` are never initialized. (I'm not entirely sure why there's no error in the view case. Perhaps the ng-init is doing some magic initialization. (By the way, I'd prefer the initialization to be done in the JS rather than through ng-init, as it can be confusing.)). Initializing both `$scope.updateRoleForm` and `$scope.viewRoleForm` to empty objects at the start of the controller should have the desired effect.